File: frontend/src/domain/tone.js
/**
 * The measured frame, in the words that go beside the numbers.
 *
 * `Shot.tone` and `Shot.motion` arrive as raw measurements — kelvin, percent,
 * frame widths — because that is what the backend stores and what the lenses
 * reason with. A readout of bare numbers teaches nothing, so the band tables
 * here name them, mirroring `app/domain/tone.py` and `app/domain/motion.py`
 * the way `guides.js` mirrors `guides.py`.
 *
 * The numbers are the point. A photographer can check 8639 K against their own
 * white balance and 1.1% clipped against their own histogram; "cool" and
 * "bright" are the adjectives this whole pass exists to replace.
 */

export const DAYLIGHT_K = 5500

const TEMPERATURE = [
  [7500, 'open shade'],
  [6500, 'overcast'],
  [5800, 'midday sun'],
  [5000, 'early or late sun'],
  [4200, 'warm interior'],
  [3200, 'tungsten'],
  [0, 'candlelight'],
]

const KEY = [
  [170, 'high key'],
  [85, 'mid key'],
  [0, 'low key'],
]

const RANGE = [
  [180, 'full range'],
  [120, 'ordinary range'],
  [0, 'flat'],
]

const PALETTE = [
  [45, 'vivid'],
  [30, 'saturated'],
  [15, 'restrained'],
  [0, 'near-neutral'],
]

function band(value, table) {
  for (const [floor, label] of table) if (value >= floor) return label
  return table[table.length - 1][1]
}

/** Nothing was measured — an old analysis, or a file Pillow could not open. */
export function measured(tone) {
  return Boolean(tone) && (tone.cct_k !== null || tone.luma_mean > 0)
}

/**
 * The readout under the frame: value, unit and the word for it. Temperature is
 * omitted rather than guessed when the frame sits too far off the Planckian
 * locus for a colour temperature to mean anything (a frame of pure red is a red
 * object, not 2655 K light), which is why `cct_k` can be null on a real shot.
 */
export function readout(tone) {
  if (!measured(tone)) return []
  const range = Math.round(tone.luma_p95 - tone.luma_p5)
  const rows = []
  if (tone.cct_k) {
    rows.push({ key: 'temperature', value: `${tone.cct_k}`, unit: 'K', label: band(tone.cct_k, TEMPERATURE) })
  }
  rows.push({ key: 'palette', value: `${Math.round(tone.saturation)}`, unit: '%', label: band(tone.saturation, PALETTE) })
  rows.push({ key: 'range', value: `${range}`, unit: '', label: band(range, RANGE) })
  rows.push({ key: 'key', value: `${Math.round(tone.luma_mean)}`, unit: '', label: band(tone.luma_mean, KEY) })
  // Only the end that actually ran out, and only when it did: a readout saying
  // "0.0% clipped white" is noise dressed as a measurement.
  if (tone.clipped_high >= 0.05) {
    rows.push({ key: 'clipped', value: tone.clipped_high.toFixed(1), unit: '%', label: 'blown white' })
  }
  if (tone.clipped_low >= 0.05) {
    rows.push({ key: 'crushed', value: tone.clipped_low.toFixed(1), unit: '%', label: 'crushed black' })
  }
  return rows
}

/** The two dominant hues and what they are doing, when there is enough colour. */
export function harmony(tone) {
  if (!measured(tone) || !tone.hues?.length || tone.saturation < 15) return ''
  const hues = tone.hues.slice(0, 2).join(' and ')
  if (tone.hue_opposition === null || tone.hue_opposition === undefined) return hues
  if (tone.hue_opposition >= 120) return `${hues}, opposed`
  if (tone.hue_opposition <= 60) return `${hues}, neighbouring`
  return hues
}

/**
 * How the camera travelled, in a phrase. Mirrors `domain/motion.py`, including
 * its silences: translation cannot see rotation, scale or focus, so a push-in
 * or an orbit reads here as "barely moves" and is left to the panel.
 */
export function cameraMove(motion) {
  if (!motion || motion.frames < 2) return ''
  const travel = Math.max(Math.abs(motion.drift_x), Math.abs(motion.drift_y))
  const seconds = motion.fps ? Math.round(motion.frames / motion.fps) : 0
  if (motion.still_share >= 0.75 && travel < 0.25) {
    return `locked off — no movement in ${Math.round(motion.still_share * 100)}% of ${seconds} s`
  }
  if (travel < 0.25) return `barely moves — ${travel.toFixed(2)} frame widths in ${seconds} s`
  const where = Math.abs(motion.drift_x) * 2 > Math.abs(motion.drift_y)
    ? motion.drift_x > 0 ? 'right' : 'left'
    : motion.drift_y > 0 ? 'down' : 'up'
  if (motion.step_max >= 0.25) {
    return `whips ${where} — ${motion.step_max.toFixed(2)} of the frame in one step, ${travel.toFixed(1)} widths in ${seconds} s`
  }
  const shape = Math.abs(motion.drift_x) > Math.abs(motion.drift_y) * 2 ? 'pans' : 'tilts'
  return `${shape} ${where} — ${travel.toFixed(1)} frame widths in ${seconds} s`
}

File: backend/app/domain/tone.py
"""Colour and tone arithmetic: what the pixels imply, computed, not guessed. Pure.

``domain/exposure.py`` sits on ``Exif`` and turns camera settings into facts.
This sits on ``Tone`` and does the same for everything the camera did not
record. It exists because the colour and light families were the two halves of
the panel making claims with nothing behind them: measured over the corpus they
fire at 0.94 and 1.06 sightings a shot, comparable to exposure's 0.94, and
until now every one of them arrived as an adjective while a composition claim
arrived with a grid reference, a guide and a percentage.

The bands are photographic convention, not invention:

* Colour temperature against the daylight locus — tungsten near 3200 K, noon
  daylight 5500 K, overcast 6500 K, open shade 7500 K and up.
* Key from mean luminance, the axis high-key and low-key actually name.
* Tonal range as the spread between the 5th and 95th percentiles, which is what
  a histogram's usable width means and what chiaroscuro pushes to the limit.

Where a band edge is a judgement rather than a convention it was placed against
the 19-frame corpus and the count it accuses is written beside it.
"""

from dataclasses import dataclass
from datetime import UTC, timedelta

from app.domain import sun
from app.domain.entities import Exif, Tone

#: Reference points on the daylight locus, warmest first.
TEMPERATURE_BANDS: tuple[tuple[int, str], ...] = (
    (7500, "open shade or a heavy blue cast"),
    (6500, "overcast daylight"),
    (5800, "midday daylight"),
    (5000, "early or late daylight"),
    (4200, "a warm interior or late golden light"),
    (3200, "tungsten"),
    (0, "candle or sodium light"),
)
#: Noon daylight, the neutral everything else is described against.
DAYLIGHT_K = 5500

#: Mean luminance on 0-255. The corpus sits between 59 and 146, so these edges
#: describe frames rather than sort this corpus: nothing here is high key.
KEY_BANDS: tuple[tuple[float, str], ...] = (
    (170, "high key"),
    (85, "mid key"),
    (0, "low key"),
)

#: Spread between the 5th and 95th percentiles of luminance.
RANGE_BANDS: tuple[tuple[float, str], ...] = (
    (180, "the full range"),
    (120, "an ordinary range"),
    (0, "a flat, compressed range"),
)

#: Mean HSV saturation as a percentage. Corpus median is 24.9 and the two ends
#: are 9.2 and 55.1, so "restrained" is the ordinary frame and both edges name
#: something a photographer would have done on purpose.
PALETTE_BANDS: tuple[tuple[float, str], ...] = (
    (45, "vivid"),
    (30, "saturated"),
    (15, "restrained"),
    (0, "near-neutral"),
)

#: Degrees apart the two dominant hues have to be to read as opposed rather
#: than as neighbours. Complementary pairs sit at 180; the corpus splits cleanly
#: into a cluster at 30 (twelve frames, all analogous) and a tail at 120-180.
COMPLEMENTARY_MIN = 120
ANALOGOUS_MAX = 60

#: Within this of sunrise or sunset the light is what photographers call golden.
#: The usual figure is the hour either side.
GOLDEN_WINDOW = timedelta(hours=1)


@dataclass(frozen=True)
class Derived:
    temperature: str
    kelvin_from_daylight: int | None
    key: str
    tonal_range: float
    range_band: str
    palette: str
    harmony: str


def band(value: float, bands: tuple[tuple[float, str], ...]) -> str:
    for floor, label in bands:
        if value >= floor:
            return label
    return bands[-1][1]


def harmony_of(tone: Tone) -> str:
    """What the two dominant hues are doing to each other."""
    if tone.saturation < PALETTE_BANDS[-2][0]:
        return "near-neutral: too little colour for a hue relationship"
    if tone.hue_opposition is None:
        return ""
    if tone.hue_opposition >= COMPLEMENTARY_MIN:
        return "opposed, the complementary relationship"
    if tone.hue_opposition <= ANALOGOUS_MAX:
        return "neighbouring, an analogous palette"
    return "a split relationship, neither opposed nor neighbouring"


def derive(tone: Tone) -> Derived:
    return Derived(
        temperature=band(tone.cct_k, TEMPERATURE_BANDS) if tone.cct_k else "",
        kelvin_from_daylight=(tone.cct_k - DAYLIGHT_K) if tone.cct_k else None,
        key=band(tone.luma_mean, KEY_BANDS),
        tonal_range=round(tone.luma_p95 - tone.luma_p5, 1),
        range_band=band(tone.luma_p95 - tone.luma_p5, RANGE_BANDS),
        palette=band(tone.saturation, PALETTE_BANDS),
        harmony=harmony_of(tone),
    )


def solar_context(exif: Exif) -> str:
    """Where the sun was when the frame was taken, when the camera recorded
    enough to say. Golden hour and blue hour are claims about the sun's
    position, so they are the two light techniques that can be checked rather
    than believed — the same NOAA equations the Scout times experiments with."""
    if exif.captured_at is None or exif.latitude is None or exif.longitude is None:
        return ""
    when = exif.captured_at
    if when.tzinfo is None:
        when = when.replace(tzinfo=UTC)
    times = sun.sun_times(when.date(), exif.latitude, exif.longitude)
    if times.polar:
        return "no sunrise or sunset at this latitude on this date"
    for label, at in (("sunrise", times.sunrise), ("sunset", times.sunset)):
        gap = when - at
        if abs(gap) <= GOLDEN_WINDOW:
            side = "after" if gap >= timedelta(0) else "before"
            return f"shot {abs(gap).seconds // 60} minutes {side} {label}: golden hour"
    if when < times.dawn or when > times.dusk:
        return "shot with the sun below civil twilight: night, not blue hour"
    if when < times.sunrise or when > times.sunset:
        return "shot between civil twilight and the horizon: blue hour"
    midday = times.sunrise + (times.sunset - times.sunrise) / 2
    hours = abs((when - midday).total_seconds()) / 3600
    return f"shot {hours:.1f} hours from solar noon, sun well up: not golden or blue hour"


def _measured(tone: Tone) -> bool:
    return tone.cct_k is not None or tone.luma_mean > 0


def technical(tone: Tone) -> list[str]:
    """For the Technician: where the frame ran out of scale. Exposure's half of
    tone, and the half a lens judging sharpness and noise should be holding."""
    if not _measured(tone):
        return []
    d = derive(tone)
    lines = [
        f"tonal range {d.tonal_range:.0f} of 255, p5 {tone.luma_p5:.0f} to "
        f"p95 {tone.luma_p95:.0f}: {d.range_band}"
    ]
    if tone.clipped_high or tone.clipped_low:
        lines.append(
            f"{tone.clipped_high:.2f}% of the frame is clipped to white and "
            f"{tone.clipped_low:.2f}% crushed to black — no detail recoverable either way"
        )
    else:
        lines.append("nothing is clipped to white or crushed to black")
    return lines


def light(tone: Tone, exif: Exif | None = None) -> list[str]:
    """For the Composer: the temperature and the key, which is what the light
    family is actually claiming about. Solar position where the camera wrote
    enough to place the sun, because golden hour and blue hour are the two
    light techniques that can be checked rather than believed."""
    if not _measured(tone):
        return []
    d = derive(tone)
    lines: list[str] = []
    if tone.cct_k:
        offset = d.kelvin_from_daylight or 0
        against = (
            f"{abs(offset)} K {'above' if offset > 0 else 'below'} daylight"
            if abs(offset) >= 300
            else "on daylight"
        )
        lines.append(f"colour temperature {tone.cct_k} K: {d.temperature} ({against})")
    lines.append(
        f"mean luminance {tone.luma_mean:.0f} of 255: {d.key}; "
        f"the spread p5 to p95 is {d.tonal_range:.0f}, {d.range_band}"
    )
    if exif is not None:
        solar = solar_context(exif)
        if solar:
            lines.append(solar)
    return lines


def palette(tone: Tone) -> list[str]:
    """For the Storyteller: how much colour there is and what it is doing.
    The colour family's evidence, which until now was an adjective."""
    if not _measured(tone):
        return []
    d = derive(tone)
    lines = [
        f"saturation {tone.saturation:.0f}% mean, {tone.saturation_p95:.0f}% at the 95th "
        f"percentile: a {d.palette} palette"
    ]
    if tone.hues:
        joined = " and ".join(tone.hues[:2])
        gap = f", {tone.hue_opposition}° apart on the wheel" if tone.hue_opposition else ""
        lines.append(f"dominant hues {joined}{gap}" + (f" — {d.harmony}" if d.harmony else ""))
    if abs(tone.cast) >= 5:
        which = "warm" if tone.cast > 0 else "cool"
        lines.append(
            f"{which} overall: mean red is {abs(tone.cast):.0f} points "
            f"{'above' if tone.cast > 0 else 'below'} mean blue on 0-255"
        )
    lines.append(
        f"{tone.warm_share:.0f}% of the frame reads warm by hue and {tone.cool_share:.0f}% cool; "
        f"{tone.accent_share:.0f}% is strongly saturated (a single accent is a small share "
        f"of an otherwise quiet frame)"
    )
    return lines


def describe(tone: Tone, exif: Exif | None = None) -> list[str]:
    """Every measured line, deduplicated. The feed, the Synthesizer and the
    Judge get the whole picture; a lens gets only its own half, so the panel
    keeps the independence that makes its vote worth taking."""
    lines: list[str] = []
    for group in (light(tone, exif), palette(tone), technical(tone)):
        for entry in group:
            if entry not in lines:
                lines.append(entry)
    return lines
